### PR TITLE
embeddable validator

### PR DESCRIPTION
**Pull request by pookmish**
# READY FOR REVIEW

# Summary
- Added a plugin manager for providing embed code validation and processing.

# Need Review By (Date)
- 9/24

# Urgency
- High

# Steps to Test
1. Follow steps in [stanford_profile PR](https://github.com/SU-SWS/stanford_profile/pull/450)

# See Also
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)

### CODE REVIEW

File: src/Plugin/EmbedValidator/LocalistValidator.php [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ */
+class LocalistValidator extends EmbedValidatorBase {
+
+  /**
+   * {@inheritDoc}
+   */
+  public function isEmbedCodeAllowed($code): bool {
+    $code = str_replace("\n", ' ', $code);
+    preg_match('/id="localist-widget-/', $code, $localist_id);
+    preg_match('/<script.*src=".*stanford.*localist.*?"/', $code, $localist_script);
```
[the remark this answers is not in the document]
Comment (pookmish) [reply src/Plugin/EmbedValidator/LocalistValidator.php:23]: correct. and basically we just create new plugins to validate those services we determine alright, with whatever logic we see fit.

File: src/Plugin/Validation/Constraint/EmbeddableConstraint.php [hunk 2]
```diff
@@ -15,6 +15,6 @@
  */
 class EmbeddableConstraint extends OEmbedResourceConstraint {
 
-  public $oEmbedNotAllowed = 'An oEmbed link is not permitted on an Embeddable which has a custom Embed Code.';
+  public $embedCodeNotAllowed = 'The given embeddable code is not permitted.';
```
Comment (imonroe) [inline src/Plugin/Validation/Constraint/EmbeddableConstraint.php:18]: I'd vote to link to a help article here for clarity, but that's just a nice-to-have.

File: src/Plugin/media/Source/Embeddable.php [hunk 3]
```diff
@@ -144,4 +176,49 @@ public function getSourceFieldValue(MediaInterface $media) {
...
+   */
+  protected function getEnabledValidationPlugins(): array {
+    $plugins = [];
+    foreach ($this->configuration['embed_validation'] as $plugin_id) {
+      /** @var \Drupal\stanford_media\Plugin\EmbedValidatorInterface $plugin */
+      try {
+        $plugins[$plugin_id] = $this->embedValidation->createInstance($plugin_id);
+      }
+      catch (\Exception $e) {
+        // The plugin didn't exist, ignore it.
```
Comment (imonroe) [inline src/Plugin/media/Source/Embeddable.php:218]: Maybe not a bad idea to log this, should it occur.

**Review by imonroe — APPROVED** [review]
This works correctly, and it's a nice bit of code.  Tests are included. A couple of small notes, but it looks like those have been addressed already.  Good to go.

**Pull request merged**